presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkDiskPageInput.java

```diff
@@ -0,0 +1,147 @@
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.spark.util.PrestoSparkUtils.computeChecksum;
+import static com.facebook.presto.spark.util.PrestoSparkUtils.toPrestoSparkSerializedPage;
+import static com.facebook.presto.spark.util.PrestoSparkUtils.toSerializedPage;
+import static java.util.Objects.requireNonNull;
+import static org.apache.spark.api.java.StorageLevels.MEMORY_AND_DISK;
+
+public class PrestoSparkDiskPageInput
+        implements PrestoSparkPageInput
+{
+    private static final Logger log = Logger.get(PrestoSparkDiskPageInput.class);
+
+    private final PagesSerde pagesSerde;
+    @GuardedBy("this")
+    private final List<Iterator<PrestoSparkSerializedPage>> serializedPageIterators;
+    @GuardedBy("this")
+    private int currentIteratorIndex;
+
+    public PrestoSparkDiskPageInput(PagesSerde pagesSerde,
+            List<List<PrestoSparkDiskPage>> broadcastTableFilesInfo,
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext)
+    {
+        this.pagesSerde = requireNonNull(pagesSerde, "pagesSerde is null");
+        this.serializedPageIterators = broadcastTableFilesInfo.stream()
+                .map(tableFiles -> loadBroadcastTableFromDisk(tableFiles, tempStorage, tempDataOperationContext))
+                .collect(Collectors.toList());
+    }
+
+    private synchronized Iterator<PrestoSparkSerializedPage> loadBroadcastTableFromDisk(
+            List<PrestoSparkDiskPage> broadcastTaskFilesInfo,
+            TempStorage tempStorage,
+            TempDataOperationContext tempDataOperationContext)
+    {
+        // Empty broadcast table
+        if (broadcastTaskFilesInfo.size() == 0) {
+            return Collections.emptyIterator();
+        }
+
+        List<PrestoSparkSerializedPage> pages = new ArrayList<>();
+        BlockManager blockManager = SparkEnv.get().blockManager();
+        BroadcastBlockId broadcastBlockId =
+                new BroadcastBlockId((broadcastTaskFilesInfo.get(0)).getBroadcastId(), "pos");
+
+        // Get the pages cached in blockmanger
+        Option<Object> object = blockManager.getSingle(broadcastBlockId, classTag(Object.class));
+        if (object.isEmpty()) {
+            try {
+                Collections.shuffle(broadcastTaskFilesInfo);
+                for (PrestoSparkTaskOutput taskFileInfo : broadcastTaskFilesInfo) {
+                    PrestoSparkDiskPage prestoSparkDiskPage = (PrestoSparkDiskPage) taskFileInfo;
+                    log.info("Reading path: " + prestoSparkDiskPage.getFilePathURI());
+                    InputStream inputStream = tempStorage.open(tempDataOperationContext, prestoSparkDiskPage.getFilePathURI());
+                    Iterator<SerializedPage> pagesIter = PagesSerdeUtil.readSerializedPages(new InputStreamSliceInput(inputStream));
```

Comment:
nit: In presto we don't use acronyms and abbreviation. Let's call it `pagesIter`. Also static import `readSerializedPages `